code/ui/blocks/src/blocks/useOf.ts

```diff
@@ -1,22 +1,51 @@
-import type { DocsContextProps } from '@storybook/types';
+import type {
+  DocsContextProps,
+  ModuleExport,
+  NormalizedProjectAnnotations,
+  PreparedMeta,
+  Renderer,
+  ResolvedModuleExportFromType,
+  ResolvedModuleExportType,
+} from '@storybook/types';
 import { prepareMeta } from '@storybook/preview-api';
 import { useContext } from 'react';
 import { DocsContext } from './DocsContext';
 
 export type Of = Parameters<DocsContextProps['resolveOf']>[0];
+
+export type EnhancedResolvedModuleExportType<
+  TType extends ResolvedModuleExportType,
+  TRenderer extends Renderer = Renderer
+> = TType extends 'component'
+  ? ResolvedModuleExportFromType<TType, TRenderer> & {
+      projectAnnotations: NormalizedProjectAnnotations<Renderer>;
+    }
+  : TType extends 'meta'
+  ? ResolvedModuleExportFromType<TType, TRenderer> & { preparedMeta: PreparedMeta }
+  : ResolvedModuleExportFromType<TType, TRenderer>;
+
 /**
  * A hook to resolve the `of` prop passed to a block.
  * will return the resolved module
  * if the resolved module is a meta it will include a preparedMeta property similar to a preparedStory
  * if the resolved module is a component it will include the project annotations
  */
-export const useOf = (...args: Parameters<DocsContextProps['resolveOf']>) => {
+export const useOf = <
+  TType extends ResolvedModuleExportType,
+  TRenderer extends Renderer = Renderer
+>(
+  moduleExportOrType: ModuleExport | TType,
+  validTypes?: TType[]
+): EnhancedResolvedModuleExportType<TType, TRenderer> => {
   const context = useContext(DocsContext);
-  const resolved = context.resolveOf(...args);
+  const resolved = context.resolveOf(moduleExportOrType, validTypes);
 
   switch (resolved.type) {
     case 'component': {
-      return { ...resolved, projectAnnotations: context.projectAnnotations };
+      return {
+        ...resolved,
+        projectAnnotations: context.projectAnnotations,
+      } as EnhancedResolvedModuleExportType<TType, TRenderer>;
```

Comment:
in reality we're just forcing the types here to be that of the return type. It could be more specific, as in "this is the return type in the 'component' case" but I couldn't figure out how to write that type, and it didn't make a difference at the call-site type anyway.

Reply:
Yeah, it doesn't need to be that precise as the caller won't know which path it went down anyway. I think the point is just "if the caller asked for component or meta" you return a `X<component or meta>`, rather than specifically `X<component>` IYKWIM.